<a href="https://colab.research.google.com/github/molaitan/Predicting-Credit-Risk/blob/main/Credit_Risk_Evaluator.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [18]:
import numpy as np
import pandas as pd
from pathlib import Path
import warnings
warnings.filterwarnings('ignore')
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

In [4]:
train_df = pd.read_csv(Path('2019loans.csv'))
test_df = pd.read_csv(Path('2020Q1loans.csv'))

In [5]:
train_df.head(2)

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0     7000.0    0.1894       256.38       MORTGAGE     75000.0   
1    40000.0    0.1614       975.71       MORTGAGE    102000.0   

  verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0        Not Verified          n  28.62          0.0             2.0  ...   
1     Source Verified          n  11.72          2.0             0.0  ...   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0              87.5                   0.0        0.0         352260.0   
1               0.0                   0.0        0.0         294664.0   

  total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0           62666.0         35000.0                     10000.0   
1          109911.0          9000.0                     71044.0   

   hardship_flag  debt_settlement_flag    target  
0              N                     N  low_risk  
1              N                     N  low_risk  

[2 rows x 84 columns]

In [6]:
test_df.head(2)

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    40000.0    0.1033       856.40           RENT    128700.0   
1    24450.0    0.1430       572.72       MORTGAGE     44574.0   

  verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0     Source Verified          n  12.47          0.0             1.0  ...   
1        Not Verified          n  15.05          0.0             1.0  ...   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0              57.1                   0.0        0.0          63915.0   
1               0.0                   0.0        0.0         136425.0   

  total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0           49510.0         49400.0                     14515.0   
1           19439.0         15500.0                     18925.0   

   hardship_flag  debt_settlement_flag    target  
0              Y                     N  low_risk  
1              N                     N  low_risk  

[2 rows x 84 columns]

In [7]:
# Convert categorical data to numeric and separate target feature for training data
x_train= pd.get_dummies(train_df.drop(columns=['target']))
y_train= train_df['target']

In [8]:
x_train.head(2)

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0     7000.0    0.1894       256.38     75000.0  28.62          0.0   
1    40000.0    0.1614       975.71    102000.0  11.72          2.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0             2.0      20.0      0.0    40414.0  ...   
1             0.0      10.0      0.0    43531.0  ...   

   verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                             0             1                      1   
1                             0             1                      0   

   initial_list_status_w  application_type_Individual  \
0                      0                            1   
1                      1                            1   

   application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                           0                1                0   
1                           0                1                0   

   debt_settlement_flag_N  debt_settlement_flag_Y  
0                       1                       0  
1                       1                       0  

[2 rows x 92 columns]

In [9]:
y_train.head(2)

0    low_risk
1    low_risk
Name: target, dtype: object

In [15]:
# Convert categorical data to numeric and separate target feature for training data
x_test= pd.get_dummies(test_df.drop(columns=['target']))
y_test= test_df['target']

In [12]:
x_test.head(2)

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0     7000.0    0.1894       256.38     75000.0  28.62          0.0   
1    40000.0    0.1614       975.71    102000.0  11.72          2.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0             2.0      20.0      0.0    40414.0  ...   
1             0.0      10.0      0.0    43531.0  ...   

   verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                             0             1                      1   
1                             0             1                      0   

   initial_list_status_w  application_type_Individual  \
0                      0                            1   
1                      1                            1   

   application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                           0                1                0   
1                           0                1                0   

   debt_settlement_flag_N  debt_settlement_flag_Y  
0                       1                       0  
1                       1                       0  

[2 rows x 92 columns]

In [13]:
y_test.head(2)

0    low_risk
1    low_risk
Name: target, dtype: object

In [16]:
# add missing dummy variables to testing set
for col in x_train.columns:
  if col not in x_test.columns:
    x_test[col]=0

Prediction on unscaled data: I predict that the Random Forest Classifier Model will perform better on unscaled data because it performs better with the several smaller trees on data with various parameters. Each tree would build from a subset of data, making the combined trees a stronger classifier. 

In [19]:
# Train the Logistic Regression model on the unscaled data and print the model score
classifier = LogisticRegression()
classifier.fit(x_train, y_train)
classifier.score(x_test, y_test)


0.5082943428328371

In [20]:
# Train a Random Forest Classifier model and print the model score
rfc= RandomForestClassifier()
rfc.fit(x_train, y_train)
rfc.score(x_test, y_test)

0.6335601871544024

The prediction that the Random Forest Classifier model would prove to perform better was proven true with a test score of 0.633 which is slight larger than the logistic regression test score of 0.508. Both scores show that both are not the strongest models because they have a 51-60% accuracy rate on the tested data. Not the best classification model to prove high or low risk loan candidates.

In [22]:
# Scale the data
from sklearn.preprocessing import StandardScaler
scaler= StandardScaler()
scaler.fit(x_train)
x_train_scaled= scaler.transform(x_train)
x_test_scaled=scaler.transform(x_test)

Prediction on scaled data: I predict that the Random Forest Classifier Model will perform better again on unscaled data because it performs better with the several smaller trees on data with various parameters. Each tree would build from a subset of data, making the combined trees a stronger classifier.

In [23]:
# Train the Logistic Regression model on the scaled data and print the model score
classifier = LogisticRegression()
classifier.fit(x_train_scaled, y_train)
classifier.score(x_test_scaled, y_test)


0.7598894087622289

In [24]:
# Train a Random Forest Classifier model and print the model score
rfc= RandomForestClassifier()
rfc.fit(x_train_scaled, y_train)
rfc.score(x_test_scaled, y_test)

0.6390897490429605

The prediction that the Random Forest Classifier model would prove to perform better was proven false with a test score of 0.639 which is less than the logistic regression test score of 0.760. Both scores show that both are not the strongest models because they have a 60-70% accuracy rate on the tested data. However, the scaled data appear to be a better set of data to produce slightly better model scores than the unscaled data models.